src/process.c

```diff
@@ -111,6 +113,153 @@ char * rcutils_get_executable_name(rcutils_allocator_t allocator)
   return executable_name;
 }
 
+rcutils_process_t *
+rcutils_start_process(
+  const rcutils_string_array_t * args,
+  rcutils_allocator_t * allocator)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(args, NULL);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(allocator, NULL);
+  if (args->size < 1) {
+    RCUTILS_SET_ERROR_MSG("args list is empty");
+    return NULL;
+  }
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    allocator, "allocator is invalid", return NULL);
+
+  rcutils_process_t * process = allocator->zero_allocate(
+    1, sizeof(rcutils_process_t), allocator->state);
+  if (NULL == process) {
+    return NULL;
+  }
+  process->allocator = *allocator;
+
+#if defined _WIN32 || defined __CYGWIN__
+  LPSTR cmd = rcutils_join(args, " ", *allocator);
+  if (NULL == cmd) {
+    rcutils_process_close(process);
+    return NULL;
+  }
+
+  STARTUPINFO start_info = {sizeof(start_info)};
+  PROCESS_INFORMATION process_info = {0};
+  if (!CreateProcess(
+          NULL, cmd, NULL, NULL, TRUE, 0,
+          NULL, NULL, &start_info, &process_info))
+  {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to create process: %lu", GetLastError());
+    allocator->deallocate(cmd, &allocator->state);
+    rcutils_process_close(process);
+    return NULL;
+  }
+
+  allocator->deallocate(cmd, &allocator->state);
+
+  CloseHandle(process_info.hThread);
+
+  process->handle = process_info.hProcess;
+  process->pid = process_info.dwProcessId;
+
+  return process;
+#else
+  rcutils_string_array_t argv = rcutils_get_zero_initialized_string_array();
+  if (RCUTILS_RET_OK != rcutils_string_array_init(&argv, args->size + 1, allocator)) {
+    return NULL;
+  }
+
+  const char * cmd = args->data[0];
+  for(size_t i = 0; i < args->size; i++) {
+    argv.data[i] = args->data[i];
+  }
+
+  process->pid = fork();
+  if (-1 == process->pid) {
+    int error = errno;
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to fork process: %d", error);
+    rcutils_process_close(process);
+    return NULL;
+  } else if (0 != process->pid) {
+    return process;
+  }
+
+  int exec_ret = execvp(cmd, argv.data);
+
+  RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING("Failed to execute process: %d", exec_ret);
+
+  for(size_t i = 0; i < argv.size; i++) {
+    argv.data[i] = NULL;
+  }
```

Comment:
i do not think we need to do this, because `rcutils_string_array_fini` does that with deallocation.

Reply:
I'm actually "borrowing" the individual strings from the passed argument array, and if they're still referenced in `argv` when I deallocate it, they will be deallocated as well.

This really shouldn't be an `rcutils_string_array_t` anyway. I think I'll switch to just a calloc/memcpy to simplify the code and make that clearer.

Reply:
Changed to a standalone zero_allocate in 7d97d42.